File: src/CourseProject/OutputOnDisplay.java
package CourseProject;

import CourseProject.mainClasses.Student;
import CourseProject.mainClasses.Teacher;
import CourseProject.mainClasses.SecondaryClasses.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Iterator;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeSet;

/**
 * Абстрактный класс, описывающий заполнение списков из базы данных и их
 * вывод на экран.
 * @author Асеев С.С.
 * @version 1.0
 */
public abstract class OutputOnDisplay
{

   /**
    * Метод для чтения из базы данных.
    * @param arg имя списка, используемого как база данных
    * @param listOfStudents список студентов
    * @param listOfTeachers список преподавателей
    */
   public static void readFromFile(String arg, TreeSet listOfStudents,
           TreeSet listOfTeachers)
   {
      try (Scanner sc = new Scanner(new File(arg), "Windows-1251"))
      {
         String str;
         while (sc.hasNext())
         {
            str = sc.nextLine();
            try (Scanner scStr = new Scanner(str))
            {
               String surname, firstname, patronymic;
               int yearOfBirthday;
               surname = scStr.next();
               firstname = scStr.next();
               patronymic = scStr.next();
               yearOfBirthday = scStr.nextInt();
               if (scStr.hasNextInt())
               {
                  int yearOfReceit = scStr.nextInt(), course = scStr.nextInt(),
                          formOfStudy = scStr.nextInt(),
                          departmentInt = scStr.nextInt();
                  boolean department = false;
                  String recordBook = scStr.next(), groupName = scStr.next();
                  StringBuilder pulpit = new StringBuilder();
                  String cipher;
                  if (departmentInt == 1)
                     department = true;
                  while (true)
                  {
                     String partOfStr = scStr.next();
                     if (partOfStr.length() >= 2 && partOfStr.charAt(2) == '.')
                     {
                        cipher = partOfStr;
                        break;
                     }
                     pulpit.append(partOfStr).append(' ');
                  }
                  String faculty = scStr.nextLine().trim();
                  listOfStudents.add(new Student(surname, firstname, patronymic,
                           yearOfBirthday, new Group(groupName,
                           new Specialization(pulpit.toString().trim(), cipher,
                           new Faculty(faculty))), course, recordBook,
                           formOfStudy, department, yearOfReceit));
               }
               else
               {
                  StringBuilder[] pulpit = new StringBuilder[2];
                  StringBuilder[] faculty = new StringBuilder[pulpit.length];
                  String[] phone = new String[pulpit.length];
                  String[] position = new String[pulpit.length];
                  int[] audience = new int[pulpit.length],
                          salary = new int[pulpit.length];
                  int i = 0;
                  while (scStr.hasNext())
                  {
                     faculty[i] = new StringBuilder();
                     pulpit[i] = new StringBuilder();
                     while (true)
                     {
                        if (scStr.hasNextInt())
                        {
                           audience[i] = scStr.nextInt();
                           phone[i] = scStr.next();
                           break;
                        }
                        String partOfStr = scStr.next();
                        pulpit[i].append(partOfStr).append(' ');
                     }
                     while (true)
                     {
                        String partOfStr;
                        if (!scStr.hasNextInt())
                        {
                           partOfStr = scStr.next();
                           if (scStr.hasNextInt())
                           {
                              position[i] = partOfStr;
                              salary[i] = scStr.nextInt();
                              break;
                           }
                           else
                           {
                              faculty[i].append(partOfStr).append(' ');
                              partOfStr = scStr.next();
                              if (scStr.hasNextInt())
                              {
                                 position[i] = partOfStr;
                                 salary[i] = scStr.nextInt();
                                 break;
                              }
                              else
                                 faculty[i].append(partOfStr).append(' ');
                           }
                        }
                     }
                     i++;
                  }
                  String[] pulpitStr = new String[pulpit.length],
                          facultyStr = new String[pulpit.length];
                  for (i = 0; i < pulpitStr.length; i++)
                  {
                     if (pulpit[i] != null)
                     {
                        pulpitStr[i] = pulpit[i].toString().trim();
                        facultyStr[i] = faculty[i].toString().trim();
                     }
                     else
                        break;
                  }
                  listOfTeachers.add(new Teacher(surname, firstname, patronymic,
                          yearOfBirthday, pulpitStr, audience, phone, facultyStr,
                          position, salary));
               }
            }
         }
      } catch (FileNotFoundException exc)
      {
         System.out.println("Файл не найден");
      }
   }

   /**
    * Метод для вывода на экран списка студентов.
    * @param listOfStudents список студентов
    */
   public static void showStudents(TreeSet listOfStudents)
   {
      Iterator<Student> itr = listOfStudents.iterator();
      int count = 1;
      printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
              + "год рождения|год поступления|курс|форма обучения|отделение|"
              + "группа|шифр специализации|# зачетной книжки|"
              + "                      факультет                      |");
      while (itr.hasNext())
      {
         Student current = itr.next();
         System.out.printf("%3d|%-15s|%-15s|%-15s|%12d|%15d|%4d|%-14s|%-9s|%6s|%18s|"
                 + "%17s|%-53s|\n", count, current.getSurname(),
                 current.getFirstname(), current.getPatronymic(),
                 current.getYearOfBirth(), current.getYearOfReceipt(),
                 current.getCourse(), current.getFormOfStudy(),
                 current.getDepartment(), current.getgName().getName(),
                 current.getgName().getsName().getCipher(), current.getRecordBook(),
                 current.getgName().getsName().getfName().getName());
         count++;
      }
   }

   /**
    * Метод для вывода на экран списка преподавателей.
    * @param listOfTeachers список преподавателей
    */
   public static void showTeachers(TreeSet listOfTeachers)
   {
      Iterator<Teacher> itr = listOfTeachers.iterator();
      int count = 1;
      printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
              + "год рождения|"
              + "                                  кафедра                     "
              + "             |"
              + "                      факультет                      "
              + "|    должность    |"
              + "                                  кафедра                     "
              + "             |"
              + "                      факультет                      "
              + "|    должность    |");
      while (itr.hasNext())
      {
         Teacher current = itr.next();
         Iterator<Map.Entry<Position, Pulpit>> itrForHM;
         itrForHM = current.getPosition().entrySet().iterator();
         System.out.printf("%3d|%-15s|%-15s|%-15s|%12d|", count,
                 current.getSurname(), current.getFirstname(),
                 current.getPatronymic(), current.getYearOfBirth());
         while (itrForHM.hasNext())
         {
            Map.Entry entry = itrForHM.next();
            System.out.printf("%-75s|%-53s|%-17s|",
                    ((Pulpit) entry.getValue()).getName(),
                    ((Pulpit) entry.getValue()).getfName().getName(),
                    ((Position) entry.getKey()).getName());
         }
         System.out.println();
         count++;
      }
   }

   /**
    * Метод для вывода списка студентов и преподавателей по определенному
    * факультету.
    * @param FACULTY список факультетов
    * @param chosenFaculty доступ к выбранному пользователем факультету
    * @param listOfStudents список студентов
    * @param listOfTeachers список преподавателей
    */
   public static void showFaculties(final String[] FACULTY, int chosenFaculty,
           TreeSet listOfStudents, TreeSet listOfTeachers)
   {
      Iterator<Student> itrStd = listOfStudents.iterator();
      Iterator<Teacher> itrTch = listOfTeachers.iterator();
      int count = 1;
      printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
              + "группа/должность");
      while (itrTch.hasNext())
      {
         boolean flag = false;
         Teacher current = itrTch.next();
         Iterator<Map.Entry<Position, Pulpit>> itrForHM;
         itrForHM = current.getPosition().entrySet().iterator();
         String position = new String();
         while (itrForHM.hasNext())
         {
            Map.Entry entry = itrForHM.next();
            if (((Pulpit) entry.getValue()).getfName().getName()
                    .equals(FACULTY[chosenFaculty]))
            {
               flag = true;
               position = ((Position) entry.getKey()).getName();
            }
            else
               flag = false;
         }
         if (flag)
         {
            System.out.printf("%3d|%-15s|%-15s|%-15s|%-16s\n", count,
                    current.getSurname(), current.getFirstname(),
                    current.getPatronymic(), position);
            count++;
         }
      }
      while (itrStd.hasNext())
      {
         Student current = itrStd.next();
         if (current.getgName().getsName().getfName().getName()
                 .equals(FACULTY[chosenFaculty]))
         {
            System.out.printf("%3d|%-15s|%-15s|%-15s|%-16s\n", count,
                    current.getSurname(), current.getFirstname(),
                    current.getPatronymic(), current.getgName().getName());
            count++;
         }
      }
   }

   /**
    * Метод для вывода на экран заголовка таблицы с ограничением снизу.
    * @param printStr заголовок
    */
   public static void printTitleWithUnderScore(String printStr)
   {
      System.out.println(printStr);
      for (int i = 0; i < printStr.length(); i++)
         System.out.print("_");
      System.out.println();
   }
}

File: OutputOnDisplay.cs
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using CourseProjectCSver.mainClasses;
using CourseProjectCSver.mainClasses.SecondaryClasses;


namespace CourseProjectCSver
{
    /**
    * Абстрактный класс, описывающий заполнение списков из базы данных и их
    * вывод на экран.
    * @author Асеев С.С.
    * @version 1.0
    */
    class OutputOnDisplay
    {

        /**
         * Метод для чтения из базы данных.
         * @param arg имя списка, используемого как база данных
         * @param listOfStudents список студентов
         * @param listOfTeachers список преподавателей
         */
        public static void readFromFile(string arg, SortedSet<Student> listOfStudents,
                SortedSet<Teacher> listOfTeachers)
        {
            try
            {
                StreamReader sc = new StreamReader(arg,
                    Encoding.GetEncoding(1251));
                string[] scStr = null;
                while (!(sc.EndOfStream))
                {
                    int index = 0;
                    string surname, firstname, patronymic;
                    int yearOfBirthday;
                    int yearOfReceit = 0;
                    scStr = sc.ReadLine().Split(';');
                    surname = scStr[index++];
                    firstname = scStr[index++];
                    patronymic = scStr[index++];
                    yearOfBirthday = int.Parse(scStr[index++]);
                    if (int.TryParse(scStr[index], out yearOfReceit))
                    {
                        int course = int.Parse(scStr[++index]),
                            formOfStudy = int.Parse(scStr[++index]),
                            departmentInt = int.Parse(scStr[++index]);
                        bool department = false;
                        string recordBook = scStr[++index],
                            groupName = scStr[++index],
                            pulpit = scStr[++index],
                            cipher = scStr[++index],
                            faculty = scStr[++index];
                        if (departmentInt == 1)
                            department = true;
                        listOfStudents.Add(new Student(surname, firstname, patronymic,
                                    yearOfBirthday, new Group(groupName,
                                    new Specialization(pulpit.ToString().Trim(), cipher,
                                    new Faculty(faculty.ToString().Trim()))), course, recordBook,
                                    formOfStudy, department, yearOfReceit));
                    }
                    else
                    {
                        string[] pulpit = new string[2],
                            faculty = new string[pulpit.Length],
                            phone = new string[pulpit.Length],
                            position = new string[pulpit.Length];
                        int[] audience = new int[pulpit.Length],
                                salary = new int[pulpit.Length];
                        for (int i = 0; i < pulpit.Length && index < scStr.Length; i++)
                        {
                            pulpit[i] = scStr[index];
                            audience[i] = int.Parse(scStr[++index]);
                            phone[i] = scStr[++index];
                            faculty[i] = scStr[++index];
                            position[i] = scStr[++index];
                            salary[i] = int.Parse(scStr[++index]);
                            index++;
                        }
                        listOfTeachers.Add(new Teacher(surname, firstname, patronymic,
                                yearOfBirthday, pulpit, audience, phone, faculty,
                                position, salary));
                    }
                }
            }
            catch (FileNotFoundException exc)
            {
                Console.WriteLine("Файл не найден");
            }
        }

            /**
             * Метод для вывода на экран списка студентов.
             * @param listOfStudents список студентов
             */
            public static void showStudents(SortedSet<Student> listOfStudents)
        {
            IEnumerator<Student> itr = listOfStudents.GetEnumerator();
            int count = 1;
            printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
                    + "год рождения|год поступления|курс|форма обучения|отделение|"
                    + "группа|шифр специализации|# зачетной книжки|"
                    + "                      факультет                      |");
            while (itr.MoveNext())
            {
                Student current = itr.Current;
                Console.WriteLine ("{0,3}|{1,-15}|{2,-15}|{3,-15}|{4,12}|" +
                    "{5,15}|{6,4}|{7,-14}|{8,-9}|{9,6}|{10,18}|{11,17}" +
                    "|{12,-53}|", count, current.getSurname(),
                        current.getFirstname(), current.getPatronymic(),
                        current.getYearOfBirth(), current.getYearOfReceipt(),
                        current.getCourse(), current.getFormOfStudy(),
                        current.getDepartment(), current.getgName().getName(),
                        current.getgName().getsName().getCipher(), current.getRecordBook(),
                        current.getgName().getsName().getfName().getName());
                count++;
            }
        }

        /**
         * Метод для вывода на экран списка преподавателей.
         * @param listOfTeachers список преподавателей
         */
        public static void showTeachers(SortedSet<Teacher> listOfTeachers)
        {
            IEnumerator<Teacher> itr = listOfTeachers.GetEnumerator();
            int count = 1;
            printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
                    + "год рождения|"
                    + "                                  кафедра                     "
                    + "             |"
                    + "                      факультет                      "
                    + "|    должность    |"
                    + "                                  кафедра                     "
                    + "             |"
                    + "                      факультет                      "
                    + "|    должность    |");
            while (itr.MoveNext())
            {
                Teacher current = itr.Current;
                IDictionaryEnumerator itrForD;
                itrForD = current.getPosition().GetEnumerator(); 
                Console.Write("{0,3}|{1,-15}|{2,-15}|{3,-15}|{4,12}|", count,
                        current.getSurname(), current.getFirstname(),
                        current.getPatronymic(), current.getYearOfBirth());
                while (itrForD.MoveNext())
                {
                    Console.Write("{0,-75}|{1,-53}|{2,-17}|",
                            ((Pulpit)itrForD.Value).getName(),
                            ((Pulpit)itrForD.Value).getfName().getName(),
                            ((Position)itrForD.Key).getName());
                    
                }
                Console.WriteLine();
                count++;
            }
        }

        /**
         * Метод для вывода списка студентов и преподавателей по определенному
         * факультету.
         * @param FACULTY список факультетов
         * @param chosenFaculty доступ к выбранному пользователем факультету
         * @param listOfStudents список студентов
         * @param listOfTeachers список преподавателей
         */
        public static void showFaculties(string[] FACULTY, int chosenFaculty,
                SortedSet<Student> listOfStudents, SortedSet<Teacher> listOfTeachers)
        {
            IEnumerator<Student> itrStd = listOfStudents.GetEnumerator();
            IEnumerator<Teacher> itrTch = listOfTeachers.GetEnumerator();
            int count = 1;
            printTitleWithUnderScore("#  |    Фамилия    |      Имя      |   Отчество    |"
                    + "группа/должность");
            while (itrTch.MoveNext())
            {
                bool flag = false;
                Teacher current = itrTch.Current;
                IDictionaryEnumerator itrForD;
                itrForD = current.getPosition().GetEnumerator();
                string position = null;
                while (itrForD.MoveNext())
                {
                    if (((Pulpit)itrForD.Value).getfName().getName()
                            .Equals(FACULTY[chosenFaculty]))
                    {
                        flag = true;
                        position = ((Position)itrForD.Key).getName();
                    }
                    else
                        flag = false;
                }
                if (flag)
                {
                    Console.WriteLine("{0,3}|{1,-15}|{2,-15}|{3,-15}|{4,-16}", count,
                            current.getSurname(), current.getFirstname(),
                            current.getPatronymic(), position);
                    count++;
                }
            }
            while (itrStd.MoveNext())
            {
                Student current = itrStd.Current;
                if (current.getgName().getsName().getfName().getName()
                        .Equals(FACULTY[chosenFaculty]))
                {
                    Console.WriteLine("{0,3}|{1,-15}|{2,-15}|{3,-15}|{4,-16}", count,
                            current.getSurname(), current.getFirstname(),
                            current.getPatronymic(), current.getgName().getName());
                    count++;
                }
            }
        }

        /**
         * Метод для вывода на экран заголовка таблицы с ограничением снизу.
         * @param printStr заголовок
         */
        public static void printTitleWithUnderScore(string printStr)
        {
            Console.WriteLine(printStr);
            for (int i = 0; i < printStr.Length; i++)
                Console.Write("_");
            Console.WriteLine();
        }
    }
}
